### PR TITLE
Bug 1913306: quick-start-footer

### PR DESCRIPTION
**Pull request by abhinandan13jan**
# Addresses
https://issues.redhat.com/browse/ODC-4619,
https://issues.redhat.com/browse/ODC-4780/'
Fixes shadow retainment on QuickStart change
Fixes shadow update on Quickstart Load
Fixes shadow update on browser resize
Markdown area scrollbar bug -> https://issues.redhat.com/browse/ODC-5338

# Fix
**Issue**
1. Unnecessary scrollbar appears in Quick start panel - This is because some headings are two lines long and some are one line long and the height computation is made by assuming one line header.
2. No shadow for overlapping content and footer.

**Solution**
1. Converted panel into flex items
2. Added shadow to footer top onScroll

# ScreenShot
Shadow update on Quickstart change and browser resize
footer1
footer2

# Tests
none changed

# Browser conformance
Chrome

### CODE REVIEW

**Comment by abhinandan13jan** [comment]
cc: @christianvogt @lwrigh

**Comment by lwrigh** [comment]
LGTM!

File: frontend/packages/console-app/src/components/quick-starts/QuickStartController.tsx [hunk 1]
```diff
@@ -39,41 +52,73 @@ const QuickStartController: React.FC<QuickStartControllerProps> = ({ quickStart
...
-  }, [previousStep]);
+    if (taskNumber > -1) return setQuickStartTaskNumber(name, taskNumber - 1);
+
+    return null;
+  }, [name, setQuickStartTaskNumber, taskNumber]);
 
   const handleTaskSelect = React.useCallback(
     (selectedTaskNumber: number) => {
       setQuickStartTaskNumber(name, selectedTaskNumber);
+      startQuickStart();
```
Comment (christianvogt) [inline frontend/packages/console-app/src/components/quick-starts/QuickStartController.tsx:94]: Looks like this is unrelated to the footer issue. Please raise a separate bug / pr for these other changes.

File: frontend/packages/console-app/src/components/quick-starts/QuickStartController.tsx [hunk 2]
```diff
@@ -39,41 +52,73 @@ const QuickStartController: React.FC<QuickStartControllerProps> = ({ quickStart
...
     },
-    [name, setQuickStartTaskNumber],
+    [name, setQuickStartTaskNumber, startQuickStart],
   );
 
+  const headerHeight = panelHeadReference.current?.offsetHeight;
+  const computedMaxHeight = window.innerHeight - (headerHeight + 6 * 24); // CSS equivalent of calc(viewportheight - (headerheight + 6 * spacer-md))
```
Comment (christianvogt) [inline frontend/packages/console-app/src/components/quick-starts/QuickStartController.tsx:100]: Do not compute the height based on window height. Always create components to fit their available height without assuming you know the parent container and its size.

A flow layout that uses sticky or flex or grid would work best.

**Review by christianvogt — COMMENTED** [review]
The scroll container should not be padded. The padding should be inside the scroll container.
The current implementation causes the shadow to appear above the scroll container and also the scroll bar is offset from the right edge:
image

**Comment by christianvogt** [comment]
@abhinandan13jan I don't see a bottom shadow.
Also if the footer and header are converted from sticky to have the scrollbar appear within the content area instead of spanning the entire height of the quick start panel, I don't know if shadows are still necessary. @lwrigh what's the UX stance on when to show these shadows?

**Comment by lwrigh** [comment]
> @abhinandan13jan I don't see a bottom shadow.
> Also if the footer and header are converted from sticky to have the scrollbar appear within the content area instead of spanning the entire height of the quick start panel, I don't know if shadows are still necessary. @lwrigh what's the UX stance on when to show these shadows?

If the scrollbar is always present inside of the quick start content area there shouldn't be a need to also have the shadows in my opinion. @bmignano What are your thoughts?

**Comment by bmignano** [comment]
> If the scrollbar is always present inside of the quick start content area there shouldn't be a need to also have the shadows in my opinion. @bmignano What are your thoughts?
I think that logic makes sense, I don't really see a need for both. But can we be sure that the scrollbar is always present / visible, even if a user's browser settings are set otherwise? @christianvogt

**Comment by christianvogt** [comment]
@lwrigh  @bmignano here are the two options and whether we still want the shadow in either or both cases. It makes some sense to have the shadow for (A) and less necessary for (B)

(A) With the entire panel scrollable and sticky header and footer:
image

(B) With a flex layout and the content container scrollable only:
image

**Comment by christianvogt** [comment]
There's confusion here. The scrollbar being present always is an OS option. We shouldn't be trying to detect when it's present and do something different.

My inquiry was related to the location of the scrollbar as indicated in my previous comment.

**Comment by lwrigh** [comment]
> @lwrigh @bmignano here are the two options and whether we still want the shadow in either or both cases. It makes some sense to have the shadow for (A) and less necessary for (B)
> 
> (A) With the entire panel scrollable and sticky header and footer:
> image
> 
> (B) With a flex layout and the content container scrollable only:
> image

Option B makes more sense to me. I definitely prefer having the scrollbar located inside of the content area and having the title header sticky.

**Comment by abhinandan13jan** [comment]
@christianvogt @lwrigh Converted to flex, moved padding inside and added shadow to footer top on scroll. Please take another look.

File: frontend/packages/console-app/src/components/quick-starts/QuickStartController.tsx (hunk 1 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/QuickStartController.tsx]: removed

File: frontend/packages/console-app/src/components/quick-starts/QuickStartController.tsx (hunk 2 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/QuickStartController.tsx]: converted to flex

**Comment by lwrigh** [comment]
@abhinandan13jan Can you post an updated screenshot for reference?

**Comment by abhinandan13jan** [comment]
@lwrigh 
> @abhinandan13jan Can you post an updated screenshot for reference?

quickstartfooterscroll

**Comment by lwrigh** [comment]
LGTM!

File: frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx [hunk 3]
```diff
@@ -47,7 +49,7 @@ const QuickStartFooter: React.FC<QuickStartFooterProps> = ({
   ]);
 
   return (
-    <div className="co-quick-start-footer">
+    <div className={footerClass}>
```
Comment (christianvogt) [inline frontend/packages/console-app/src/components/quick-starts/controller/QuickStartFooter.tsx:52]: `co-quick-start-footer` should always be present. Do not rely on the parent to pass in this class.
Concatenate the incoming class to this.

File: frontend/packages/console-app/src/components/quick-starts/QuickStartPanelContent.scss [hunk 4]
```diff
@@ -10,6 +10,10 @@
     background: inherit;
     z-index: var(--pf-global--ZIndex--xs);
   }
+  &-body {
```
Comment (christianvogt) [inline frontend/packages/console-app/src/components/quick-starts/QuickStartPanelContent.scss:13]:
```suggestion
  &__body {
```

File: frontend/packages/console-app/src/components/quick-starts/controller/QuickStartContent.scss [hunk 5]
```diff
@@ -1,7 +1,7 @@
 @import '~bootstrap-sass/assets/stylesheets/bootstrap/variables';
 
 .co-quick-start-content {
-  @media (min-width: $screen-lg-min) {
-    min-height: calc(100vh - (var(--pf-global--spacer--4xl) + var(--pf-global--spacer--3xl)));
-  }
+  flex: 1 1 0;
+  overflow: auto;
+  padding: var(--pf-global--spacer--sm) var(--pf-global--spacer--lg);;
```
Comment (christianvogt) [inline frontend/packages/console-app/src/components/quick-starts/controller/QuickStartContent.scss:6]:
```suggestion
  padding: var(--pf-global--spacer--sm) var(--pf-global--spacer--lg);
```

**Review by christianvogt — COMMENTED** [review]
If you start with no scroll bar and then reduce the vertical space of the browser, you end up with scrollbars appearing inside the markdown content:
image

If you scroll such that a shadow appears, then change quick starts such that there is no scrollbar present, the previous shadow remains. This is also visible in the GIF you posted where the bottom shadow does not disappear.
image

File: frontend/packages/console-app/src/components/quick-starts/controller/QuickStartContent.scss (hunk 5 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/controller/QuickStartContent.scss]: updated

File: frontend/packages/console-app/src/components/quick-starts/QuickStartPanelContent.scss (hunk 4 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/QuickStartPanelContent.scss]: updated

**Comment by abhinandan13jan** [comment]
> If you start with no scroll bar and then reduce the vertical space of the browser, you end up with scrollbars appearing inside the markdown content:
> image
> 
> If you scroll such that a shadow appears, then change quick starts such that there is no scrollbar present, the previous shadow remains. This is also visible in the GIF you posted where the bottom shadow does not disappear.
> image

@christianvogt These are existing issues. The one with markup scroll is because of hardcoded styles in the markUp component.
The scroll issue is also there on the cluster. I tried to work around with that, but need to update the hook a little which may become a breaking change.

**Comment by christianvogt** [comment]
@abhinandan13jan please log a bug for the markdown scrollbar appearing and please fix the shadows so they correctly disappear otherwise it's misleading.

**Comment by abhinandan13jan** [comment]
> @abhinandan13jan please log a bug for the markdown scrollbar appearing and please fix the shadows so they correctly disappear otherwise it's misleading.

Fixed the shadow issue and for the markup logged https://issues.redhat.com/browse/ODC-5338

shadow-issue-fix

**Comment by sahil143** [comment]
@abhinandan13jan I'm not seeing the shadow on the footer unless I scroll. The shadow should be present if the content is scrollable to indicate that there is content present off the screen.
Screenshot 2021-01-13 at 7 07 27 PM

**Comment by sahil143** [comment]
The header shadow doesn't seem to be fixed.
quick-start-scroll

**Comment by abhinandan13jan** [comment]
Maybe we need to redo the QuickStart SidePanel to achieve what we are looking for. But I believe, both the issues mentioned in the above JIRA tickets have been addressed. 
cc: @andrewballantyne @nimishamukherjee @christianvogt

**Comment by abhinandan13jan** [comment]
@sahil143 I updated the PR. could you please re-review?

File: frontend/packages/console-app/src/components/quick-starts/controller/QuickStartContent.tsx [hunk 6]
```diff
@@ -13,52 +14,74 @@ type QuickStartContentProps = {
...
+
+    const debouncedResize = useDebounceCallback(resizeCallback, [resizeCallback], 50);
+
+    React.useEffect(() => {
+      const resizeObserver: ResizeObserver = new ResizeObserver(debouncedResize);
+      resizeObserver.observe(ref.current);
+      return () => {
+        resizeObserver.disconnect();
+      };
+    }, [ref, onContentUpdate, debouncedResize]);
```
Comment (sahil143) [inline frontend/packages/console-app/src/components/quick-starts/controller/QuickStartContent.tsx:51]: use the hook here  `useResizeObserver`

Comment (sahil143) [inline frontend/packages/console-app/src/components/quick-starts/controller/QuickStartContent.tsx:51]: Why is `onContentUpdate` part of dependency array?

File: frontend/packages/console-app/src/components/quick-starts/controller/QuickStartContent.tsx [hunk 7]
```diff
@@ -13,52 +14,74 @@ type QuickStartContentProps = {
...
+    } = props;
+    const {
+      spec: { introduction, tasks, conclusion, nextQuickStart = [] },
+    } = quickStart;
+    const totalTasks = tasks.length;
+    const nextQS = nextQuickStart.length > 0 && nextQuickStart[0];
+
+    const resizeCallback: ResizeCallback = () => {
+      ref?.current && onContentUpdate();
+    };
```
Comment (sahil143) [inline frontend/packages/console-app/src/components/quick-starts/controller/QuickStartContent.tsx:41]: seems like you are calling this on resizeObserver only. Why do you need to forwardRef here? Why not just use the hook `useRef` and pass the current value to `onContentUpdate(ref.current)`?

File: frontend/packages/console-app/src/components/quick-starts/QuickStartPanelContent.tsx [hunk 8]
```diff
@@ -27,18 +37,40 @@ const QuickStartPanelContent: React.FC<QuickStartPanelContentProps> = ({
   handleClose,
   activeQuickStartID,
 }) => {
-  const [scrollDirection, handleScrollCallback] = useScrollDirection();
+  const [scrollDirection, handleScrollCallback, resetScrollDirection] = useScrollDirection();
   const { t } = useTranslation();
   const quickStart = quickStarts.find((qs) => qs.metadata.name === activeQuickStartID);
 
-  const headerClasses = classNames('co-quick-start-panel-content-head', {
+  const updateShadow: UpdateShadow = (contentRef) => {
```
Comment (sahil143) [inline frontend/packages/console-app/src/components/quick-starts/QuickStartPanelContent.tsx:44]: use `useCallback` here

File: frontend/packages/console-app/src/components/quick-starts/controller/QuickStartContent.tsx [hunk 9]
```diff
@@ -13,52 +14,74 @@ type QuickStartContentProps = {
...
+      onContentUpdate,
+    } = props;
+    const {
+      spec: { introduction, tasks, conclusion, nextQuickStart = [] },
+    } = quickStart;
+    const totalTasks = tasks.length;
+    const nextQS = nextQuickStart.length > 0 && nextQuickStart[0];
+
+    const resizeCallback: ResizeCallback = () => {
+      ref?.current && onContentUpdate();
```
Comment (sahil143) [inline frontend/packages/console-app/src/components/quick-starts/controller/QuickStartContent.tsx:40]:
```suggestion
      ref.current && onContentUpdate();
```

**Review by sahil143 — CHANGES_REQUESTED** [review]

File: frontend/packages/console-app/src/components/quick-starts/controller/QuickStartContent.tsx (hunk 6 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/controller/QuickStartContent.tsx:51]: it's a lint error otherwise

Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/controller/QuickStartContent.tsx:51]: I don't need the width or the height. It is all about checking the scrollHeight vs clientHeight. Can you please tell me how the `onResizeObserver` would serve any better, given we have a more custom implementation of the same mechanism above??

File: frontend/packages/console-app/src/components/quick-starts/controller/QuickStartContent.tsx (hunk 7 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/controller/QuickStartContent.tsx]: Updated. thanks for pointing it out.

File: frontend/packages/console-app/src/components/quick-starts/QuickStartPanelContent.tsx (hunk 8 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/QuickStartPanelContent.tsx]: updated. Using a memoised callback now.

File: frontend/packages/console-app/src/components/quick-starts/controller/QuickStartContent.tsx (hunk 9 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/controller/QuickStartContent.tsx]: updated

File: frontend/packages/console-app/src/components/quick-starts/controller/QuickStartContent.tsx (hunk 6 above)
Comment (sahil143) [reply frontend/packages/console-app/src/components/quick-starts/controller/QuickStartContent.tsx]: replace this useEffect with
```
useResizeObserver(debouncedResize, ref.current)
```

Comment (sahil143) [reply frontend/packages/console-app/src/components/quick-starts/controller/QuickStartContent.tsx]: This is custom hook. you might need to rebase with master.

Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/controller/QuickStartContent.tsx]: updated this as well

**Comment by christianvogt** [comment]
The logic for computing the shadows is incorrect.

You need an `onScroll` event listener on the content scroll container. There is no need for a resize observer. Ideally the parent component which is managing the shadows should simply pass along an `onScroll` event handler which gets attached to the content scroll container div.

This use to work in this way when the drawer itself was the scroll container. You still have the event handler attached to the `DrawerPanelContent` component as `onScroll={handleScrollCallback}` even though this component no longer scrolls with the current implementation.

**Comment by abhinandan13jan** [comment]
@sahil143 @christianvogt Updated with the resizescroller hook.Please check

File: frontend/packages/console-shared/src/hooks/scroll.ts [hunk 10]
```diff
@@ -45,6 +53,41 @@ export const useScrollDirection = (): [ScrollDirection, (event) => void] => {
     },
     [scrollDirection],
   );
-
   return [scrollDirection, handleScroll];
 };
+
+export const useScrollShadows = (node: HTMLElement, debounce = 200): Shadows => {
```
Comment (sahil143) [inline frontend/packages/console-shared/src/hooks/scroll.ts:59]: move this to a separate file

Comment (sahil143) [inline frontend/packages/console-shared/src/hooks/scroll.ts:59]:
```suggestion
export const useScrollShadows = (node: HTMLElement, debounce: number = 200): Shadows => {
```

File: frontend/packages/console-shared/src/hooks/scroll.ts [hunk 11]
```diff
@@ -1,4 +1,5 @@
 import * as React from 'react';
+import { useResizeObserver, useDebounceCallback } from '@console/shared';
```
Comment (sahil143) [inline frontend/packages/console-shared/src/hooks/scroll.ts:2]: use relative path to import these
```suggestion
import { useResizeObserver } from './useResizeObserver';
import { useDebounceCallback } from './debounce';
```

File: frontend/packages/console-app/src/components/quick-starts/QuickStartPanelContent.tsx [hunk 12]
```diff
@@ -27,18 +29,19 @@ const QuickStartPanelContent: React.FC<QuickStartPanelContentProps> = ({
   handleClose,
   activeQuickStartID,
 }) => {
-  const [scrollDirection, handleScrollCallback] = useScrollDirection();
   const { t } = useTranslation();
+  const [shadows, SetShadows] = React.useState<Shadows>(Shadows.none);
```
Comment (sahil143) [inline frontend/packages/console-app/src/components/quick-starts/QuickStartPanelContent.tsx:33]: use camelCasing for names
```suggestion
  const [shadows, setShadows] = React.useState<Shadows>(Shadows.none);
```

File: frontend/packages/console-app/src/components/quick-starts/QuickStartPanelContent.tsx [hunk 13]
```diff
@@ -27,18 +29,19 @@ const QuickStartPanelContent: React.FC<QuickStartPanelContentProps> = ({
...
   const quickStart = quickStarts.find((qs) => qs.metadata.name === activeQuickStartID);
 
-  const headerClasses = classNames('co-quick-start-panel-content-head', {
-    'pf-u-box-shadow-sm-bottom':
-      scrollDirection && scrollDirection !== ScrollDirection.scrolledToTop,
+  const updateShadow: UpdateShadow = (shadow) => SetShadows(shadow);
```
Comment (sahil143) [inline frontend/packages/console-app/src/components/quick-starts/QuickStartPanelContent.tsx:36]: You can directly pass `setShadow` to the component. Why do you need to wrap it?

File: frontend/packages/console-app/src/components/quick-starts/QuickStartController.tsx [hunk 14]
```diff
@@ -28,6 +37,22 @@ const QuickStartController: React.FC<QuickStartControllerProps> = ({ quickStart
     (task, index) => activeQuickStartState[`taskStatus${index}`],
   ) as QuickStartTaskStatus[];
 
+  const [contentRef, SetContentRef] = React.useState<HTMLDivElement>(null);
```
Comment (sahil143) [inline frontend/packages/console-app/src/components/quick-starts/QuickStartController.tsx:40]: use camelCasing 
```suggestion
  const [contentRef, setContentRef] = React.useState<HTMLDivElement>(null);
```

File: frontend/packages/console-app/src/components/quick-starts/QuickStartController.tsx [hunk 15]
```diff
@@ -28,6 +37,22 @@ const QuickStartController: React.FC<QuickStartControllerProps> = ({ quickStart
...
+  const ref = React.createRef<HTMLDivElement>();
+  const shadows = useScrollShadows(contentRef, 100);
```
Comment (sahil143) [inline frontend/packages/console-app/src/components/quick-starts/QuickStartController.tsx:42]: Why not pass ref.current directly here?
```suggestion
  const shadows = useScrollShadows(ref.current, 100);
```

Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/QuickStartController.tsx:42]: The hook will not update unless ref.current is set as a state. This is where `contentRef` comes in and stores ref.current as state

File: frontend/packages/console-app/src/components/quick-starts/QuickStartController.tsx (hunk 14 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/QuickStartController.tsx]: updated

File: frontend/packages/console-app/src/components/quick-starts/QuickStartPanelContent.tsx (hunk 13 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/QuickStartPanelContent.tsx]: updated

File: frontend/packages/console-app/src/components/quick-starts/QuickStartPanelContent.tsx (hunk 12 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/QuickStartPanelContent.tsx]: updated

File: frontend/packages/console-shared/src/hooks/scroll.ts (hunk 11 above)
Comment (abhinandan13jan) [reply frontend/packages/console-shared/src/hooks/scroll.ts]: updated

File: frontend/packages/console-app/src/components/quick-starts/QuickStartController.tsx (hunk 15 above)
Comment (christianvogt) [reply frontend/packages/console-app/src/components/quick-starts/QuickStartController.tsx:39]: Can you move this hook to `QuickStartPanelContent.tsx` and add a prop `contentRef` that allows you to pass a ref function down to the get to the content. That way `QuickStartPanelContent.tsx` can manage the the shadows.

File: frontend/packages/console-app/src/components/quick-starts/QuickStartController.tsx [hunk 16]
```diff
@@ -28,6 +34,22 @@ const QuickStartController: React.FC<QuickStartControllerProps> = ({ quickStart
...
+
+  React.useEffect(() => {
+    ref.current !== contentRef && setContentRef(ref.current);
+  }, [ref, contentRef]);
+
+  React.useEffect(() => {
+    setShadows(shadows);
+  }, [shadows, setShadows]);
```
Comment (christianvogt) [inline frontend/packages/console-app/src/components/quick-starts/QuickStartController.tsx:47]: You won't need these hooks if you move the `useScrollShadows` to `QuickStartPanelContent`. Furthermore the first `useEffect` is invalid because `ref` is a constant and would never update even if the `ref.current` value changed.

File: frontend/packages/console-shared/src/hooks/scrollShadows.ts [hunk 17]
```diff
@@ -0,0 +1,46 @@
...
+export enum Shadows {
+  none = 'none',
+  both = 'both',
+  top = 'top',
+  bottom = 'bottom',
+}
+
+export const useScrollShadows = (node: HTMLElement, debounce: number = 200): Shadows => {
+  const [shadows, setShadows] = React.useState(Shadows.none);
+  const computeShadows = useDebounceCallback(() => {
```
[the remark this answers is not in the document]
Comment (abhinandan13jan) [reply frontend/packages/console-shared/src/hooks/scrollShadows.ts]: removed Debouncing

File: frontend/packages/console-shared/src/hooks/scrollShadows.ts [hunk 18]
```diff
@@ -0,0 +1,46 @@
...
+import { useDebounceCallback } from './debounce';
+
+export enum Shadows {
+  none = 'none',
+  both = 'both',
+  top = 'top',
+  bottom = 'bottom',
+}
+
+export const useScrollShadows = (node: HTMLElement, debounce: number = 200): Shadows => {
```
[the remark this answers is not in the document]
Comment (abhinandan13jan) [reply frontend/packages/console-shared/src/hooks/scrollShadows.ts]: file renamed

File: frontend/packages/console-app/src/components/quick-starts/QuickStartController.tsx (hunk 16 above)
Comment (abhinandan13jan) [reply frontend/packages/console-app/src/components/quick-starts/QuickStartController.tsx]: updated

File: frontend/packages/console-app/src/components/quick-starts/QuickStartController.tsx [hunk 19]
```diff
@@ -6,9 +6,15 @@ import { QuickStartContext, QuickStartContextValues } from './utils/quick-start-
 
 type QuickStartControllerProps = {
   quickStart: QuickStart;
+  footerClass: string;
+  updateContentRef: (ref: HTMLDivElement) => void;
```
Comment (christianvogt) [inline frontend/packages/console-app/src/components/quick-starts/QuickStartController.tsx:10]: Rename and update type:
```suggestion
  contentRef: React.Ref<HTMLDivElement>;
```

File: frontend/packages/console-app/src/components/quick-starts/QuickStartController.tsx [hunk 20]
```diff
@@ -28,6 +34,9 @@ const QuickStartController: React.FC<QuickStartControllerProps> = ({ quickStart
     (task, index) => activeQuickStartState[`taskStatus${index}`],
   ) as QuickStartTaskStatus[];
 
+  const ref = React.createRef<HTMLDivElement>();
+  React.useEffect(() => updateContentRef(ref?.current));
```
Comment (christianvogt) [inline frontend/packages/console-app/src/components/quick-starts/QuickStartController.tsx:38]:
```suggestion
```

This approach is incorrect because an update to the `ref` doesn't cause a re-render and therefore the effect won't get run unless something else triggers the re-render. You need to use a function as the ref.

But with my other suggestions, you won't need this.

File: frontend/packages/console-app/src/components/quick-starts/QuickStartController.tsx [hunk 21]
```diff
@@ -65,13 +74,15 @@ const QuickStartController: React.FC<QuickStartControllerProps> = ({ quickStart
         onTaskSelect={handleTaskSelect}
         onTaskReview={handleTaskStatusChange}
         onQuickStartChange={handleQuickStartChange}
+        ref={ref}
```
Comment (christianvogt) [inline frontend/packages/console-app/src/components/quick-starts/QuickStartController.tsx:77]: Pass the incoming prop straight down to the content child:
```suggestion
        ref={contentRef}
```

File: frontend/packages/console-app/src/components/quick-starts/QuickStartPanelContent.tsx [hunk 22]
```diff
@@ -10,12 +10,14 @@ import {
   Title,
 } from '@patternfly/react-core';
 import { AsyncComponent } from '@console/internal/components/utils';
-import { useScrollDirection, ScrollDirection } from '@console/shared/';
+import { useScrollShadows, Shadows } from '@console/shared/';
```
Comment (christianvogt) [inline frontend/packages/console-app/src/components/quick-starts/QuickStartPanelContent.tsx:13]:
```suggestion
import { useScrollShadows, Shadows } from '@console/shared';
```

File: frontend/packages/console-app/src/components/quick-starts/QuickStartPanelContent.tsx [hunk 23]
```diff
@@ -10,12 +10,14 @@ import {
...
 import { QuickStart } from './utils/quick-start-types';
 import './QuickStartPanelContent.scss';
 
 type HandleClose = () => void;
 
+type UpdateContentRef = (newRef: HTMLDivElement) => void;
```
Comment (christianvogt) [inline frontend/packages/console-app/src/components/quick-starts/QuickStartPanelContent.tsx:19]:
```suggestion
```

File: frontend/packages/console-app/src/components/quick-starts/QuickStartPanelContent.tsx [hunk 24]
```diff
@@ -27,18 +29,24 @@ const QuickStartPanelContent: React.FC<QuickStartPanelContentProps> = ({
   handleClose,
   activeQuickStartID,
 }) => {
-  const [scrollDirection, handleScrollCallback] = useScrollDirection();
   const { t } = useTranslation();
+  const [contentRef, setContentRef] = React.useState<HTMLDivElement>();
+  const shadows = useScrollShadows(contentRef);
+
+  const updateContentRef: UpdateContentRef = (newRef) => setContentRef(newRef);
```
Comment (christianvogt) [inline frontend/packages/console-app/src/components/quick-starts/QuickStartPanelContent.tsx:36]:
```suggestion
```

File: frontend/packages/console-app/src/components/quick-starts/QuickStartPanelContent.tsx [hunk 25]
```diff
@@ -59,10 +67,12 @@ const QuickStartPanelContent: React.FC<QuickStartPanelContentProps> = ({
           </DrawerActions>
         </DrawerHead>
       </div>
-      <DrawerPanelBody>
+      <DrawerPanelBody hasNoPadding className="co-quick-start-panel-content__body">
         <AsyncComponent
           loader={() => import('./QuickStartController').then((c) => c.default)}
           quickStart={quickStart}
+          footerClass={footerClass}
+          updateContentRef={updateContentRef}
```
Comment (christianvogt) [inline frontend/packages/console-app/src/components/quick-starts/QuickStartPanelContent.tsx:75]:
```suggestion
          contentRef={setContentRef}
```
